src/CLI/usbguard-check-IPC-permission.cpp

```diff
@@ -0,0 +1,103 @@
+//
+// Copyright (C) 2020 Red Hat, Inc.
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation; either version 2 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+// Authors: Attila Lakatos <alakatos@redhat.com>
+//
+#ifdef HAVE_BUILD_CONFIG_H
+  #include <build-config.h>
+#endif
+
+#include "usbguard.hpp"
+#include "usbguard-check-IPC-permission.hpp"
+
+#include "usbguard/IPCClient.hpp"
+
+#include <iostream>
+
+namespace usbguard
+{
+  static const char* options_short = "hs:p:";
+
+  static const struct ::option options_long[] = {
+    { "help", no_argument, nullptr, 'h' },
+    { "section", required_argument, nullptr, 's' },
+    { "privilege", required_argument, nullptr, 'p' },
+    { nullptr, 0, nullptr, 0 }
+  };
+
+  static void showHelp(std::ostream& stream)
+  {
+    stream << " Usage: " << usbguard_arg0 << " check-IPC-permission [OPTIONS]" << std::endl;
+    stream << std::endl;
+    stream << " Options:" << std::endl;
+    stream << "  -s, --section    Specify section." << std::endl;
+    stream << "  -p, --privilege  Specify privilege." << std::endl;
+    stream << "  -h, --help       Show this help." << std::endl;
+    stream << std::endl;
+  }
+
+  int usbguard_check_IPC_permission(int argc, char* argv[])
+  {
+    std::string section_string = "";
+    std::string privilege_string = "";
+    int opt = 0;
+
+    while ((opt = getopt_long(argc, argv, options_short, options_long, nullptr)) != -1) {
+      switch (opt) {
+      case 'h':
+        showHelp(std::cout);
+        return EXIT_SUCCESS;
+
+      case 's':
+        section_string = optarg;
+        break;
+
+      case 'p':
+        privilege_string = optarg;
+        break;
+
+      case '?':
+        showHelp(std::cerr);
+
+      default:
+        return EXIT_FAILURE;
+      }
+    }
+
+    try {
+      IPCServer::AccessControl::Section section = IPCServer::AccessControl::sectionFromString(section_string);;
+      IPCServer::AccessControl::Privilege privilege = IPCServer::AccessControl::privilegeFromString(privilege_string);;
+      usbguard::IPCClient ipc(/*connected=*/true);
```

Comment:
Afaik IPCClient needs to be able to connect to IPC interface when its initialized. Is it really a good approach that you need IPC connection to check IPC permission? Maybe it would be better to have permissions for each operation stored outside IPCServer so the user would not require IPC connection to check permissions.

Reply:
The idea sounds good to me but there is a problem with that: only IPCServer knows the IPC permissions, so in order to get them it is inevitable to use IPC connection. One possible way to do that without using the IPCClient would be to read it from configuration files but it would require root privileges, so I think it's not an option to consider.